system/jlib/jlog.cpp

```diff
@@ -1538,6 +1592,43 @@ CLogMsgManager::~CLogMsgManager()
     }
 }
 
+LogMsgJobId CLogMsgManager::addJobId(const char *job)
+{
+    LogMsgJobId ret = ++nextJobId;
+    pushMsg(new LogMsg(ret, job));
+    return ret;
+}
+
+void CLogMsgManager::removeJobId(LogMsgJobId id)
+{
+    pushMsg(new LogMsg(id, nullptr));
+}
+
+const char * CLogMsgManager::queryJobId(LogMsgJobId id) const
+{
+    // NOTE - thread safety is important here. We have to consider two things:
+    // 1. Whether an id (and therefore an entry in this table) can be invalidated between the return statement and someone using the result
+    //    It is up to the calling application to ensure that it does not call removeJobId() on an ID that may still be being used for logging by another thread.
+    // 2. Whether the table lookup may coincide with a table add, and crash in getValue/setValue
+    //    This is a non-issue in queueing mode as all gets/sets happen on a single thread
```

Comment:
There is potentially a problem though if queryJobId() is called from any thread other than the logging thread.  I wonder if it should always lock..

Reply:
Depends whether the expense is a concern. I did wonder about a variant of CLeavableCriticalBlock that will assert that all accesses are from the same thread - but it's not easy to code without assuming a singleton